sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1139,14 +1141,130 @@ private[lf] object SBuiltin {
+    }
+  }
+
+  private[this] def castAnyInterface[Q](
+      machine: Machine[_],
+      ifaceId: TypeConName,
+      coid: V.ContractId,
+      actualTmplId: TypeConName,
+      record: SRecord,
+  )(k: SValue => Control[Q]): Control[Q] =
+    if (!interfaceInstanceExists(machine, ifaceId, actualTmplId)) {
+      Control.Error(IE.ContractDoesNotImplementInterface(ifaceId, coid, actualTmplId))
+    } else {
+      k(record)
+    }
+
+  /** Fetches the requested contract Id, upgrades it to the preferred template version for the same package name,
+    * and compares the computed views according to the old and the new versions. If the two views agree then caches
+    * the upgraded contract, inserts a Fetch node, and returns the upgraded contract.
+    */
+  final case class SBFetchInterface(interfaceId: TypeConName) extends UpdateBuiltin(1) {
+    override protected def executeUpdate(
+        args: util.ArrayList[SValue],
+        machine: UpdateMachine,
+    ): Control[Question.Update] = {
+      val coid = getSContractId(args, 0)
+      fetchAny(machine, None, coid, SValue.SValue.None) { (maybePkgName, sourceContract) =>
+        maybePkgName match {
+          case None =>
+            crash(s"unexpected contract instance without packageName")
+          case Some(pkgName) =>
+            val (sourceTplId, sourceArg) = getSAnyContract(ArrayList.single(sourceContract), 0)
+            castAnyInterface(machine, interfaceId, coid, sourceTplId, sourceArg) { _ =>
+              viewInterface(machine, interfaceId, sourceTplId, sourceArg) { sourceView =>
+                resolvePackageName(machine, pkgName) { pkgId =>
+                  val dstTmplId = sourceTplId.copy(packageId = pkgId)
+                  machine.ensurePackageIsLoaded(
+                    dstTmplId.packageId,
+                    language.Reference.Template(dstTmplId),
+                  ) { () =>
+                    ensureTemplateImplementsInterface(machine, dstTmplId, interfaceId) { () =>
+                      fromInterface(machine, sourceTplId, sourceArg, dstTmplId) {
+                        case None => crash(s"failed to convert contract $coid to type $dstTmplId")
+                        case Some(dstArg) =>
+                          viewInterface(machine, interfaceId, dstTmplId, dstArg) { dstView =>
+                            executeExpression(machine, sourceView) { sourceViewValue =>
+                              // If the destination and source templates are the same, we skip the computation
+                              // of the destination template's view.
+                              if (dstTmplId == sourceTplId)
+                                cacheContractAndInsertNode(machine, coid, dstTmplId, dstArg)
+                              else
+                                executeExpression(machine, dstView) { dstViewValue =>
+                                  if (sourceViewValue != dstViewValue) {
+                                    crash("view mistmatch")
```

Comment:
We need a proper error.

Reply:
Used a Dev error for now, will discuss with @tudor-da how to introduce a new ledger API error.